alf/trainers/policy_trainer.py

```diff
@@ -714,16 +718,31 @@ def play(root_dir,
 
         episode_reward += time_step_reward
 
+        # start defer encoding
+        if defer_mode and (episode_length > future_steps):
+            _record_future_info(
+                recorder,
+                observations,
+                actions,
+                rewards,
+                future_steps=future_steps)
+            observations.pop(0)
+            actions.pop(0)
+            rewards.pop(0)
+
         if time_step.is_last() or episode_length >= max_episode_length > 0:
             logging.info("episode_length=%s episode_reward=%s" %
                          (episode_length, episode_reward))
+
+            # defer encoding all the rest till the end
             if defer_mode:
-                _record_episodic_info(
+                _record_future_info(
                     recorder,
                     observations,
                     actions,
                     rewards,
-                    future_steps=future_steps)
+                    future_steps=future_steps,
+                    encode_all=True)
```

Comment:
Why do you need this encode_all?

Reply:
Need to encode what's left in the episode buffer at the end fo an episode.

Reply:
> Need to encode what's left in the episode buffer at the end fo an episode.

I think what's left in the buffer is always the length of future_steps. So it will always be encoded at the previous step.